trino/dbapi.py

```diff
@@ -228,10 +268,31 @@ def cursor(self, legacy_primitive_types: bool = None):
         return Cursor(
             self,
             request,
-            # if legacy_primitive_types is not explicitly set in Cursor, take from Connection
+            # if legacy params are not explicitly set in Cursor, take them from Connection
             legacy_primitive_types if legacy_primitive_types is not None else self.legacy_primitive_types
         )
 
+    def _use_legacy_prepared_statements(self):
+        if self.legacy_prepared_statements is not None:
+            return self.legacy_prepared_statements
+
+        value = mustUseLegacyPreparedStatements.get((self.host, self.port))
+        if value is None:
+            try:
+                query = trino.client.TrinoQuery(
+                    self._create_request(),
+                    # The version() function was introduced in Trino version 352
+                    query="SELECT node_version FROM system.runtime.nodes WHERE coordinator = true")
+                rows = query.execute().rows
+                if rows:
+                    version = rows[0][0]
+                    value = (version <= "417")
+                    mustUseLegacyPreparedStatements.put((self.host, self.port), value)
+            except Exception:
+                # not updating the cache
+                value = True
```

Comment:
in case of failure with the query we'd not cache the failure and keep retrying on every new connection creation.

Is the failure expected to be transient? IMO we should cache failure as well and revert to legacy statements in such cases - there's no downside to doing so.

Reply:
Yeah, I was thinking the failure would be transient, but your suggestion makes sense.

Reply:
Given this, what do you think about changing the query used for checking the capability.  Instead of relying on the version check, just see if the new statement works.  Basically run `EXECUTE IMMEDIATE 'SELECT 1'` and see if it succeeds.

Reply:
It makes sense to me.

I also see that SELECT can be constrained via resource groups/access control, I'm not sure if `EXECUTE` has such restrictions as well.